README.md
# enenra

An Erlang/OTP library for interfacing with Google Cloud Storage. Named after the [smoke monster](https://en.wikipedia.org/wiki/Enenra) of Japanese folklore, for no particular reason. This library provides basic CRUD operations on buckets and objects, including uploading and downloading of object media.

## Requirements

* [Erlang/OTP](http://www.erlang.org) R18 or R19
    * R20 has backward incompatible API changes
* [rebar3](https://github.com/erlang/rebar3/) 3.0.0 or higher

## Building and Testing

```
$ rebar3 compile
$ export GOOGLE_APPLICATION_CREDENTIALS=~/.your_gcp_credentials.json
$ rebar3 ct
```

Note that the test suite expects to find an environment variable named `GOOGLE_APPLICATION_CREDENTIALS` which specifies the path to your Google Cloud Platform service credentials. This JSON formatted file is created via the Cloud Console, as described in the setup section below. This file contains your private key, so be sure to store this file with permissions that prevent exposure to third parties.

## Example Usage

To include `enenra` as a dependency in your release, add it to the list of dependencies in your `rebar.config` file, like so:

```
{deps, [
    {enenra, {git, "https://github.com/nlfiedler/enenra", {tag, "0.3.0"}}}
]}.
```

To have the `enenra` application started automatically, be sure to include `enenra` in the `applications` list of your application configuration before building a release. You may also want to add `jiffy` to the list of `included_applications`, and `hackney` to the list of `applications`.

Below is a simple example in which a bucket is created and a file is uploaded to that bucket. This also demonstrates loading the credentials from a file and computing an MD5 checksum of a file to be uploaded to a bucket.

```erlang
1> rr("include/enenra.hrl").
[access_token,bucket,credentials,object]
2> application:ensure_all_started(enenra).
{ok,[idna,mimerl,certifi,ssl_verify_fun,metrics,hackney,enenra]}
3> Credentials = os:getenv("GOOGLE_APPLICATION_CREDENTIALS").
"/Users/nfiedler/.gcloud/testing.json"
4> {ok, Creds} = enenra:load_credentials(Credentials).
{ok,#credentials{type = <<"service_account">>,
                 project_id = <<"a-project">>,
                 private_key_id = <<"a-private-key-id">>,
                 private_key = <<"a-private-key">>,
                 client_email = <<"an-email-address">>,
                 client_id = <<"a-client-id">>}}
5> enenra:insert_bucket(#bucket{
    name = <<"0136d00f-a942-11e6-8f9a-3c07547e18a6-enenra-1234">>,
    location = <<"US-WEST1">>,
    storageClass = <<"STANDARD">>}, Creds).
{ok,#bucket{id = <<"a-bucket-id">>,
            projectNumber = <<"a-project-number">>,
            name = <<"0136d00f-a942-11e6-8f9a-3c07547e18a6-enenra-1234">>,
            timeCreated = <<"2016-11-18T22:25:54.239Z">>,
            updated = <<"2016-11-18T22:25:54.239Z">>,
            location = <<"US-WEST1">>,
            storageClass = <<"STANDARD">>}}
6> {ok, Md5} = enenra:compute_md5("test/enenra_SUITE_data/IMG_5745.JPG").
{ok,<<"kq56YDAH2p4mzAqrQw84kQ==">>}
7> enenra:upload_file("test/enenra_SUITE_data/IMG_5745.JPG", #object{
    name = <<"my_image">>,
    bucket = <<"0136d00f-a942-11e6-8f9a-3c07547e18a6-enenra-1234">>,
    contentType = <<"image/jpeg">>,
    md5Hash = Md5,
    size = 107302}, Creds).
{ok,#object{id = <<"a-bucket-id/name/an-object-id">>,
            name = <<"my_image">>,
            bucket = <<"0136d00f-a942-11e6-8f9a-3c07547e18a6-enenra-1234">>,
            contentType = <<"image/jpeg">>,
            timeCreated = <<"2016-11-18T22:28:01.232Z">>,
            updated = <<"2016-11-18T22:28:01.232Z">>,
            storageClass = <<"STANDARD">>,
            size = <<"107302">>,
            md5Hash = <<"kq56YDAH2p4mzAqrQw84kQ==">>}}
```

## Google Cloud Setup

1. Visit https://console.cloud.google.com/ and log in with your account.
1. Select an existing project or create a new one.
1. Note the *Project ID* as that will be necessary when connecting via enenra.
1. Using the menu, select **API Manager**, then **Credentials**.
1. On the **Credentials** page, select the **Create credentials** drop-down, then select **Service account key**.
1. From the **Service** account drop-down, select an existing service account or create a new one.
1. For **Key** type, select the **JSON** key option, then select **Create**. The file automatically downloads to your computer.
1. Put the `*.json` file you just downloaded in a directory of your choosing. This directory must be private, but accessible to your application.

## Security

The credentials file must be readable by the application, but should not be readable by casual users. Likewise, it is possible, if the server process crashes, that the private keys will end up in the log file (they are an argument to the API, after all). As such, the log files should be protected from unintended exposure to third parties.

## Docker

[Docker](https://www.docker.com) can be used to build and test the code without affecting your development environment, which may have a different version of Erlang/OTP installed. The use of `docker-compose`, as shown in the example below, is optional, but it makes the process very simple. Note that `GOOGLE_APPLICATION_CREDENTIALS` should have a path reachable within the container, such as `/src/creds.json`

```shell
$ cd docker
$ docker-compose build
$ docker-compose run enenra
$ rebar3 clean
$ rebar3 compile
$ rebar3 ct
```

## License

[BSD 3-Clause](https://opensource.org/licenses/BSD-3-Clause), see the `LICENSE` file.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.4.0] - 2020-04-11
### Changed
- kenan-gillet: Allow parallelization of calls to GCS

## [0.3.1] - 2020-04-10
### Changed
- kenan-gillet: update hackney to fix crash in erlang >= 22.1
- kenan-gillet: allow getting token from the google internal metadata server
- kenan-gillet: jiffy can just be added as an application

## [0.3.0] - 2019-02-02
### Added
- chrisplim: Add `upload_data/3` to upload an object from data
- nfiedler: Update hackney to 1.15.0 release

## [0.2.1] - 2018-11-19
### Changed
- chrisplim: Update types in hrl to pass dialyzer check.

## [0.2.0] - 2018-11-13
### Added
- pselden: Add `get_object_contents/3` to fetch an object as a binary.
### Changed
- Update hackney and jiffy to latest released versions.

## [0.1.7] - 2017-07-17
### Changed
- Update hackney version; change all HTTP request headers to binaries.
- Return HTTP 400 response as an error, rather than crashing.

## [0.1.6] - 2017-01-30
### Changed
- Raise `recv_timeout` to 5 minutes as it repeatedly timed out for a small file.

## [0.1.5] - 2017-01-20
### Changed
- Catch errors raised by `jiffy` for malformed JSON and report as an error.

## [0.1.4] - 2016-12-23
### Changed
- Return any file upload errors gracefully rather than crashing the gen_server.

## [0.1.3] - 2016-12-22
### Changed
- Ensure response body is consistently read so connection can be released.

## [0.1.2] - 2016-12-20
### Changed
- Set timeout for receiving response after a file upload to 60 seconds.

## [0.1.1] - 2016-12-18
### Changed
- Increase timeouts of `gen_server:call` to `infinity` since remote calls can be slow.

## [0.1.0] - 2016-11-18
### Changed
- Initial release


src/enenra_server.erl
%% -*- coding: utf-8 -*-
%%
%% Copyright 2016-2017 Nathan Fiedler. All rights reserved.
%% Use of this source code is governed by a BSD-style
%% license that can be found in the LICENSE file.
%%
-module(enenra_server).

-behavior(gen_server).
-export([start_link/0, call/1]).
-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).

-define(BASE_URL, <<"https://www.googleapis.com/storage/v1/b/">>).
-define(UPLOAD_URL, <<"https://www.googleapis.com/upload/storage/v1/b/">>).
-define(AUTH_URL, <<"https://www.googleapis.com/oauth2/v4/token">>).
-define(AUD_URL, <<"https://www.googleapis.com/oauth2/v4/token">>).

-define(GOOGLE_INTERNAL_AUTH_URL, <<"http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token">>).

% read/write is useful for adding and deleting, but that's about it
-define(READ_WRITE_SCOPE, <<"https://www.googleapis.com/auth/devstorage.read_write">>).
% full-control is required for updating/patching existing resources
-define(FULL_CONTROL_SCOPE, <<"https://www.googleapis.com/auth/devstorage.full_control">>).

% Base64 encoding of JSON {"alg":"RS256","typ":"JWT"}
-define(JWT_HEADER, <<"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9">>).
-define(GRANT_TYPE, <<"urn:ietf:params:oauth:grant-type:jwt-bearer">>).

-include_lib("public_key/include/public_key.hrl").
-include("enenra.hrl").

-record(state, {token}).

%%
%% Client API
%%
start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

% @doc
% direct call, only blocking on getting token

call(Msg) ->
    case msg_to_fun(Msg) of
        {Credentials, Fun} ->
            {ok, Token} = call_remote({get_token, Credentials}),
            %% Invoke the given function with an authorization token. If the function
            case Fun(Token) of
                {error, auth_required} ->
                    %% re-try after refreshing the token
                    {ok, Token} = call_remote({refresh_token, Credentials}),
                    Fun(Token);
                Result ->
                    Result
            end;
        skip -> unsupported
    end.

%%
%% gen_server callbacks
%%
init([]) ->
    {ok, #state{}}.

handle_call({refresh_token, Credentials}, _From, State) ->
    case get_auth_token(Credentials) of
      R = {ok, Token} ->
        {reply, R, State#state{token=Token}};
      Reply ->
        {reply, Reply, State#state{token=undefined}}
    end;
handle_call({get_token, Credentials}, From, State = #state{token=undefined}) ->
    % token is not set, just call refresh
    handle_call({refresh_token, Credentials}, From, State);
handle_call({get_token, _Credentials}, _From, State = #state{token=Token}) ->
    {reply, {ok, Token}, State}.

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info(_Msg, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

%%
%% Private functions
%%

call_remote(Msg) ->
  gen_server:call(?MODULE, Msg, infinity).

-define (CRED_FUN(FunBody),
    { Credentials, fun(Token) -> FunBody end }
).
%% transform message into function callbacks
msg_to_fun({list_buckets, Credentials}) ->
    ?CRED_FUN(list_buckets(Credentials, Token));
msg_to_fun({get_bucket, Name, Credentials}) ->
    ?CRED_FUN(get_bucket(Name, Token));
msg_to_fun({insert_bucket, Bucket, Credentials}) ->
    ?CRED_FUN(insert_bucket(Bucket, Credentials, Token));
msg_to_fun({update_bucket, Name, Properties, Credentials}) ->
    ?CRED_FUN(update_bucket(Name, Properties, Token));
msg_to_fun({delete_bucket, Name, Credentials}) ->
    ?CRED_FUN(delete_bucket(Name, Token));
msg_to_fun({list_objects, BucketName, Credentials}) ->
    ?CRED_FUN(list_objects(BucketName, Token));
msg_to_fun({upload_object, Object, RequestBody, Credentials}) ->
    ?CRED_FUN(upload_object(Object, RequestBody, Token));
msg_to_fun({download_object, BucketName, ObjectName, Filename, Credentials}) ->
    ?CRED_FUN(download_object(BucketName, ObjectName, Filename, Token));
msg_to_fun({get_object, BucketName, ObjectName, Credentials}) ->
    ?CRED_FUN(get_object(BucketName, ObjectName, Token));
msg_to_fun({get_object_contents, BucketName, ObjectName, Credentials}) ->
    ?CRED_FUN(get_object_contents(BucketName, ObjectName, Token));
msg_to_fun({delete_object, BucketName, ObjectName, Credentials}) ->
    ?CRED_FUN(delete_object(BucketName, ObjectName, Token));
msg_to_fun({update_object, BucketName, ObjectName, Properties, Credentials}) ->
    ?CRED_FUN(update_object(BucketName, ObjectName, Properties, Token));
msg_to_fun(_) ->
    skip.

% @doc
%
% Retrieve a list of the available buckets.
%
-spec list_buckets(credentials(), access_token()) -> {ok, [bucket()]} | {error, term()}.
list_buckets(Credentials, Token) ->
    Project = Credentials#credentials.project_id,
    Url = hackney_url:make_url(?BASE_URL, <<"">>, [{"project", Project}]),
    ReqHeaders = add_auth_header(Token, []),
    {ok, Status, Headers, Client} = hackney:request(get, Url, ReqHeaders),
    case decode_response(Status, Headers, Client) of
        {ok, Body} ->
            Items = proplists:get_value(<<"items">>, Body, []),
            {ok, [make_bucket(Item) || {Item} <- Items]};
        R -> R
    end.

% @doc
%
% Retrieve the named bucket.
%
-spec get_bucket(binary(), access_token()) -> {ok, bucket()} | {error, term()}.
get_bucket(Name, Token) ->
    Url = <<?BASE_URL/binary, Name/binary>>,
    ReqHeaders = add_auth_header(Token, []),
    {ok, Status, Headers, Client} = hackney:request(get, Url, ReqHeaders),
    case decode_response(Status, Headers, Client) of
        {ok, Body} -> {ok, make_bucket(Body)};
        R -> R
    end.

% @doc
%
% Insert a bucket with the given properties. Only name, location, and class
% are used (and are required). If the bucket already exists, the existing
% properties are retrieved and returned.
%
-spec insert_bucket(bucket(), credentials(), access_token()) -> {ok, bucket()} | {error, term()}.
insert_bucket(Bucket, Credentials, Token) ->
    Project = Credentials#credentials.project_id,
    Url = hackney_url:make_url(?BASE_URL, <<"">>, [{"project", Project}]),
    ReqHeaders = add_auth_header(Token, [
        {<<"Content-Type">>, <<"application/json">>}
    ]),
    ReqBody = binary_to_list(jiffy:encode({[
        {<<"name">>, Bucket#bucket.name},
        {<<"location">>, Bucket#bucket.location},
        {<<"storageClass">>, Bucket#bucket.storageClass}
    ]})),
    {ok, Status, Headers, Client} = hackney:request(post, Url, ReqHeaders, ReqBody),
    case decode_response(Status, Headers, Client) of
        {ok, Body} -> {ok, make_bucket(Body)};
        {error, conflict} -> get_bucket(Bucket#bucket.name, Token);
        R -> R
    end.

% @doc
%
% Update an existing bucket with the properties defined in the given
% property list (uses the PATCH method to update only those fields). The
% names and values should be binary instead of string type. To clear an
% existing field, set the field value to 'null'.
%
-spec update_bucket(binary(), list(), access_token()) -> {ok, bucket()} | {error, term()}.
update_bucket(Name, Properties, Token) ->
    Url = <<?BASE_URL/binary, Name/binary>>,
    ReqHeaders = add_auth_header(Token, [
        {<<"Content-Type">>, <<"application/json">>}
    ]),
    ReqBody = binary_to_list(jiffy:encode({Properties})),
    {ok, Status, Headers, Client} = hackney:request(patch, Url, ReqHeaders, ReqBody),
    case decode_response(Status, Headers, Client) of
        {ok, Body} -> {ok, make_bucket(Body)};
        R -> R
    end.

% @doc
%
% Deletes the named bucket. Returns ok upon success, or {error, Reason} if
% an error occurred.
%
-spec delete_bucket(binary(), access_token()) -> {ok, string()} | {error, term()}.
delete_bucket(Name, Token) ->
    Url = <<?BASE_URL/binary, Name/binary>>,
    ReqHeaders = add_auth_header(Token, []),
    {ok, Status, Headers, Client} = hackney:request(delete, Url, ReqHeaders),
    decode_response(Status, Headers, Client).

% @doc
%
% Construct a bucket record from the given property list.
%
-spec make_bucket(list()) -> bucket().
make_bucket(PropList) ->
    #bucket{
        id=proplists:get_value(<<"id">>, PropList),
        projectNumber=proplists:get_value(<<"projectNumber">>, PropList),
        name=proplists:get_value(<<"name">>, PropList),
        timeCreated=proplists:get_value(<<"timeCreated">>, PropList),
        updated=proplists:get_value(<<"updated">>, PropList),
        location=proplists:get_value(<<"location">>, PropList),
        storageClass=proplists:get_value(<<"storageClass">>, PropList)
    }.

% @doc
%
% Retrieve a list of the objects stored in the named bucket.
%
-spec list_objects(binary(), access_token()) -> {ok, object()} | {error, term()}.
list_objects(BucketName, Token) ->
    Url = <<?BASE_URL/binary, BucketName/binary, "/o">>,
    ReqHeaders = add_auth_header(Token, []),
    {ok, Status, Headers, Client} = hackney:request(get, Url, ReqHeaders),
    case decode_response(Status, Headers, Client) of
        {ok, Body} ->
            Items = proplists:get_value(<<"items">>, Body, []),
            {ok, [make_object(Item) || {Item} <- Items]};
        R -> R
    end.

% @doc
%
% Upload an object to the named bucket and return {ok, Object} or {error,
% Reason}, where Object has the updated object properties.
%
-spec upload_object(object(), request_body(), access_token()) -> {ok, object()} | {error, term()}.
upload_object(Object, RequestBody, Token) ->
    BucketName = Object#object.bucket,
    Url = hackney_url:make_url(
        ?UPLOAD_URL, <<BucketName/binary, "/o">>, [{"uploadType", "resumable"}]),
    ReqHeaders = add_auth_header(Token, [
        {<<"Content-Type">>, <<"application/json; charset=UTF-8">>},
        {<<"X-Upload-Content-Type">>, Object#object.contentType},
        {<<"X-Upload-Content-Length">>, Object#object.size}
    ]),
    ReqBody = binary_to_list(jiffy:encode({[
        {<<"name">>, Object#object.name},
        % include the md5 so GCP can verify the upload was successful
        {<<"md5Hash">>, Object#object.md5Hash}
    ]})),
    {ok, Status, Headers, Client} = hackney:request(post, Url, ReqHeaders, ReqBody),
    case Status of
        200 ->
            % need to read/skip the body to close the connection
            hackney:skip_body(Client),
            UploadUrl = proplists:get_value(<<"Location">>, Headers),
            do_upload(UploadUrl, Object, RequestBody, Token);
        _ -> decode_response(Status, Headers, Client)
    end.

% @doc
%
% Perform put request to upload the object to the given upload URL and return {ok, Object} or
% {error, Reason}, where Object has the updated object properties.
%
-spec do_upload(binary(), object(), request_body(), access_token()) -> {ok, object()} | {error, term()}.
do_upload(Url, Object, RequestBody, Token) ->
    ReqHeaders = add_auth_header(Token, [
        {<<"Content-Type">>, Object#object.contentType},
        {<<"Content-Length">>, Object#object.size}
    ]),
    % Receiving the response after an upload can take a few seconds, so
    % give it a chance to compute the MD5 and such before timing out. Set
    % the timeout rather high as it seems that certain inputs can cause a
    % long delay in response?
    Options = [{recv_timeout, 300000}],
    % Errors during upload are not unusual, so return them gracefully
    % rather than exploding and generating a lengthy crash report.

    %
    % TODO: works on Erlang 19? but not on Erlang 20?
    %
    case hackney:request(put, Url, ReqHeaders, RequestBody, Options) of
        {ok, Status, Headers, Client} ->
            case decode_response(Status, Headers, Client) of
                {ok, Body} -> {ok, make_object(Body)};
                R0 -> R0
            end;
        R1 -> R1
    end.

% @doc
%
% Retrieve the object and save to the named file.
%
-spec download_object(binary(), binary(), string(), credentials()) -> ok | {error, term()}.
download_object(BucketName, ObjectName, Filename, Token) ->
    ON = hackney_url:urlencode(ObjectName),
    UrlPath = <<BucketName/binary, "/o/", ON/binary>>,
    Url = hackney_url:make_url(?BASE_URL, UrlPath, [{"alt", "media"}]),
    ReqHeaders = add_auth_header(Token, []),
    {ok, Status, Headers, Client} = hackney:request(get, Url, ReqHeaders),
    case Status of
        200 ->
            {ok, FileHandle} = file:open(Filename, [write]),
            stream_to_file(FileHandle, Client);
        _ -> decode_response(Status, Headers, Client)
    end.

% @doc
%
% Stream the response body of the HTTP request to the opened file. Returns
% ok if successful, or {error, Reason} if not. The file will closed upon
% successful download.
%
-spec stream_to_file(term(), term()) -> ok | {error, term()}.
stream_to_file(FileHandle, Client) ->
    case hackney:stream_body(Client) of
        done -> file:close(FileHandle);
        {ok, Bin} ->
            ok = file:write(FileHandle, Bin),
            stream_to_file(FileHandle, Client);
        R -> R
    end.

% @doc
%
% Retrieve the properties of the named object in the named bucket.
%
-spec get_object(binary(), binary(), credentials()) -> {ok, object()} | {error, term()}.
get_object(BucketName, ObjectName, Token) ->
    ON = hackney_url:urlencode(ObjectName),
    Url = <<?BASE_URL/binary, BucketName/binary, "/o/", ON/binary>>,
    ReqHeaders = add_auth_header(Token, []),
    {ok, Status, Headers, Client} = hackney:request(get, Url, ReqHeaders),
    case decode_response(Status, Headers, Client) of
        {ok, Body} -> {ok, make_object(Body)};
        R -> R
    end.
% @doc
%
% Retrieve the contents of the named object in the named bucket.
%
-spec get_object_contents(binary(), binary(), credentials()) -> {ok, object()} | {error, term()}.
get_object_contents(BucketName, ObjectName, Token) ->
    ON = hackney_url:urlencode(ObjectName),
    UrlPath = <<BucketName/binary, "/o/", ON/binary>>,
    Url = hackney_url:make_url(?BASE_URL, UrlPath, [{"alt", "media"}]),
    ReqHeaders = add_auth_header(Token, []),
    {ok, Status, Headers, Client} = hackney:request(get, Url, ReqHeaders),
    case Status of
        200 -> stream_to_binary(Client);
        _ -> decode_response(Status, Headers, Client)
    end.

stream_to_binary(Client) -> stream_to_binary(Client, <<>>).
stream_to_binary(Client, Acc) ->
    case hackney:stream_body(Client) of
        done -> {ok, Acc};
        {ok, Data} -> stream_to_binary(Client, << Acc/binary, Data/binary >>);
        R -> R
    end.

% @doc
%
% Delete the named object in the named bucket.
%
-spec delete_object(binary(), binary(), credentials()) -> ok | {error, term()}.
delete_object(BucketName, ObjectName, Token) ->
    ON = hackney_url:urlencode(ObjectName),
    Url = <<?BASE_URL/binary, BucketName/binary, "/o/", ON/binary>>,
    ReqHeaders = add_auth_header(Token, []),
    {ok, Status, Headers, Client} = hackney:request(delete, Url, ReqHeaders),
    decode_response(Status, Headers, Client).

% @doc
%
% Update an existing object with the properties defined in the given
% property list (uses the PATCH method to update only those fields). The
% names and values should be binary instead of string type. To clear an
% existing field, set the field value to 'null'.
%
-spec update_object(binary(), binary(), list(), access_token()) -> {ok, object()} | {error, term()}.
update_object(BucketName, ObjectName, Properties, Token) ->
    ON = hackney_url:urlencode(ObjectName),
    Url = <<?BASE_URL/binary, BucketName/binary, "/o/", ON/binary>>,
    ReqHeaders = add_auth_header(Token, [
        {<<"Content-Type">>, <<"application/json">>}
    ]),
    ReqBody = binary_to_list(jiffy:encode({Properties})),
    {ok, Status, Headers, Client} = hackney:request(patch, Url, ReqHeaders, ReqBody),
    case decode_response(Status, Headers, Client) of
        {ok, Body} -> {ok, make_object(Body)};
        R -> R
    end.

% @doc
%
% Construct an object record from the given property list.
%
-spec make_object(list()) -> object().
make_object(PropList) ->
    #object{
        id=proplists:get_value(<<"id">>, PropList),
        name=proplists:get_value(<<"name">>, PropList),
        bucket=proplists:get_value(<<"bucket">>, PropList),
        contentType=proplists:get_value(<<"contentType">>, PropList),
        timeCreated=proplists:get_value(<<"timeCreated">>, PropList),
        updated=proplists:get_value(<<"updated">>, PropList),
        storageClass=proplists:get_value(<<"storageClass">>, PropList),
        size=proplists:get_value(<<"size">>, PropList),
        md5Hash=proplists:get_value(<<"md5Hash">>, PropList)
    }.

% @doc
%
% Add the "Authorization" header to those given and return the new list.
%
-spec add_auth_header(access_token(), list()) -> list().
add_auth_header(Token, Headers) ->
    AccessToken = Token#access_token.access_token,
    TokenType = Token#access_token.token_type,
    Authorization = <<TokenType/binary, " ", AccessToken/binary>>,
    Headers ++ [
        {<<"Authorization">>, Authorization}
    ].

% @doc
%
% Based on the response, return {ok, Body}, ok, or {error, Reason}. The
% body is the decoded JSON response from the server. The error
% 'auth_required' indicates a new authorization token must be retrieved. A
% 204 returns 'ok', while a 403 returns {error, forbidden}, 404 returns
% {error, not_found}, 409 returns {error, conflict}.
%
-spec decode_response(integer(), list(), term()) -> {ok, term()} | {error, term()}.
decode_response(400, _Headers, Client) ->
    {ok, Body} = hackney:body(Client),
    {error, Body};
decode_response(401, _Headers, Client) ->
    % need to read/skip the body to close the connection
    hackney:skip_body(Client),
    {error, auth_required};
decode_response(403, _Headers, Client) ->
    % need to read/skip the body to close the connection
    hackney:skip_body(Client),
    {error, forbidden};
decode_response(404, _Headers, Client) ->
    % need to read/skip the body to close the connection
    hackney:skip_body(Client),
    {error, not_found};
decode_response(409, _Headers, Client) ->
    % need to read/skip the body to close the connection
    hackney:skip_body(Client),
    {error, conflict};
decode_response(Ok, _Headers, Client) when Ok == 200; Ok == 201 ->
    {ok, Body} = hackney:body(Client),
    try jiffy:decode(Body) of
        {Results} -> {ok, Results}
    catch
        Error -> Error
    end;
decode_response(204, _Headers, Client) ->
    % need to read/skip the body to close the connection
    hackney:skip_body(Client),
    ok;
decode_response(_Status, _Headers, Client) ->
    {ok, Body} = hackney:body(Client),
    try jiffy:decode(Body) of
        {Results} -> {ok, Results}
    catch
        Error -> Error
    end.

% @doc
%
% Retrieve a read/write authorization token from the remote service, based
% on the provided credentials, which contains the client email address and
% the PEM encoded private key.
%
-spec get_auth_token(credentials()) -> {ok, access_token()} | {error, term()}.
get_auth_token(use_google_internal_metadata_server) ->
    %% it assumes that workload identity is properly configured
    {ok, Status, Headers, Client} = hackney:request(get, ?GOOGLE_INTERNAL_AUTH_URL, [{<<"Metadata-Flavor">>, <<"Google">>}]),
    decode_token_response(Status, Headers, Client);
get_auth_token(Creds) ->
    Now = seconds_since_epoch(),
    % GCP seems to completely ignore the timeout value and always expires
    % in 3600 seconds anyway. Who knows, maybe someday it will work, but
    % you can forget about automated testing of expiration for now.
    Timeout = application:get_env(enenra, auth_timeout, 3600),
    ClaimSet = base64:encode(jiffy:encode({[
        {<<"iss">>, Creds#credentials.client_email},
        {<<"scope">>, ?FULL_CONTROL_SCOPE},
        {<<"aud">>, ?AUD_URL},
        {<<"exp">>, Now + Timeout},
        {<<"iat">>, Now}
    ]})),
    JwtPrefix = <<?JWT_HEADER/binary, ".", ClaimSet/binary>>,
    PrivateKey = Creds#credentials.private_key,
    Signature = compute_signature(PrivateKey, JwtPrefix),
    Jwt = <<JwtPrefix/binary, ".", Signature/binary>>,
    ReqBody = {form, [{<<"grant_type">>, ?GRANT_TYPE}, {<<"assertion">>, Jwt}]},
    {ok, Status, Headers, Client} = hackney:request(post, ?AUTH_URL, [], ReqBody),
    decode_token_response(Status, Headers, Client).

decode_token_response(Status, Headers, Client) ->
    case decode_response(Status, Headers, Client) of
        {ok, Token} ->
            AccessToken = proplists:get_value(<<"access_token">>, Token),
            TokenType = proplists:get_value(<<"token_type">>, Token),
            {ok, #access_token{access_token=AccessToken, token_type=TokenType}};
        R -> R
    end.

% @doc
%
% Return the seconds since the epoch (1970/1/1 00:00).
%
-spec seconds_since_epoch() -> integer().
seconds_since_epoch() ->
    {Mega, Sec, _Micro} = os:timestamp(),
    Mega * 1000000 + Sec.

% @doc
%
% Compute the SHA256 signature of the given Msg data, using the PEM key
% binary data (which will be decoded and from which the RSA private key
% is extracted). The result is Base64 encoded, appropriate for making
% an authorization request.
%
-spec compute_signature(binary(), binary()) -> binary().
compute_signature(PemKeyBin, Msg) ->
    [PemKeyData] = public_key:pem_decode(PemKeyBin),
    PemKey = public_key:pem_entry_decode(PemKeyData),
    base64:encode(public_key:sign(Msg, sha256, PemKey)).
